src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerObserver.cs

```diff
@@ -40,7 +42,18 @@ public Task OpenAsync(ChangeFeedObserverContext context)
 
         public Task ProcessChangesAsync(ChangeFeedObserverContext context, IReadOnlyList<Document> docs)
         {
-            return this.executor.TryExecuteAsync(new TriggeredFunctionData() { TriggerValue = docs }, CancellationToken.None);
+            // Not sure why these methods aren't declared with the "async" keyword, but wrapping this logic in a task taking an async 
+            // lambda so I can use it properly and avoid blocking with a Task.Wait() or Task.Result() call.
+            return Task.Run(async () =>
+            {
+                var result = await this.executor.TryExecuteAsync(new TriggeredFunctionData() { TriggerValue = docs }, CancellationToken.None);
+
+                if (!result.Succeeded && this.haltOnFailure == true)
+                {
+                    // unfortunate that we can't re-throw and preserve stack trace here
+                    throw result.Exception;
```

Comment:
Rather than throwing, can we swap to doing the checkpointing manually here, based on this `haltOnFailure`? Could we provide more control that way?

@ealsur -- you probably know more about what's possible here...

Reply:
For example, could we do any of these things?:
- Set a max retry for a document (is there any tracking for that?)
- Set some retry delay (is there a way to 'pause' the change feed?)

Reply:
More questions :-)
- This will restart the observer -- but how long will it wait to restart? A few seconds? Minutes? Is it configurable?
- I'm assuming the entire batch is retried? So if some were successfully processed, those will be sent again as well?

Reply:
Comments on 1)  This is desired behavior for customers who require reliable processing, but I agree it should be documented well enough for them to understand the implications.  I've tried to make the comments on the attribute very descriptive, but MSDN/etc should also be updated.  Dead lettering would be a possible alternative for some but not all situations.  For example, in our production system in which failure to process is an issue that require immediate attention, I set retry count to infinite on our QueueAttributes.  A dead-lettered notification/queue item means dropped customer data and downstream failures.  Additionally, in this case, we require in-order processing as well, so we can't just check the dead letter queue periodically.  I can see it being a valid solution for other use cases though.

Comments on 2)  Suspect this new observer will be created based on lease expiry time.  The "gotcha" about "first run / no checkpoint was ever created" I believe is a general issue with the change feed library and something I've actually hit in real-life using that library directly!  My account has 24 partitions, and for a long time I couldn't figure out why some notifications may be lost if I simply shut down the web job, updated some documents, and restarted the web job.  The notifications that went to a partition that was never checkpointed just got dropped!  Understanding that this is because the checkpoint is in the form of a continuation token from DocDB it finally made sense, but was quite confusing at first.  I'm considering submitting separate feedback on this as its a problem that is not specific to CosmosDBTrigger.

Yes, it does require the function to be idempotent and have the ability to process replays of sequences instead of individual notifications.  This could be fixed by setting max batch size to 1 and checkpoint-every to 1 as well, at the cost of performance.  I don't believe these change feed processor settings are currently exposed by CosmosDBTrigger.

Reply:
(the "max batch size" and "checkpoint-every" settings I referred to are in ChangeFeedHostOptions.CheckpointFrequency)

Reply:
I agree with the concerns around infinite loops and would prefer a design along the lines of what Brett eluded to with x retries and y back off. It is already the case that customers have to be aware they are getting an "at least once" guarantee and that they need to write their code to allow for the possibility of retries so I'm not that concerned about a design that makes retries more frequent.

Can we talk more about how straightforward it would be to implement "retry x times with y back off"? I don't know the code well enough to understand whether this is something we can easily do purely within functions or if we need functionality from the change feed processor.

Reply:
Are you suggesting "before propagating the throw back to the change feed processor library and thus causing it not to checkpoint, do a couple retries under the attribute's control while holding onto the partition lock"?  I almost implemented that as part of the change, it was only a couple more lines, but given that retry will happen anyway, didn't consider it important and didn't want to confuse the pull request.  I'm curious about your thoughts here and what concerns you have that "attribute controlled retries" would resolve?  

In my situation, the most common case would be, for example, a Cosmos outage that could last from a few minutes to a few hours and I'd prefer to halt processing /ordered/ notifications until it's back up.  Or a similar issue surrounding a Splunk dependency.  (For me, in each case the write action is effectively idempotent, and even a series of retried writes would be idempotent.)

Reply:
@nkinnan Max Batch Size is coming as a new attribute too among other things, working on it here https://github.com/Azure/azure-webjobs-sdk-extensions/pull/337

Reply:
lets move this conversation out of the inline comments and into the main PR comments because they are hidden by github automatically as the code is updated.